contrib/epee/src/readline_buffer.cpp

```diff
@@ -196,15 +199,50 @@ static int startup_hook()
   return 0;
 }
 
+static char* completion_matches(const char* text, int state)
+{
+  static size_t list_index;
+  static int len;
+
+  if(state == 0)
+  {
+    list_index = 0;
+    len = strlen(text);
+  }
+
+  const std::vector<std::string>& completions = rdln::readline_buffer::get_completions();
+  for(; list_index<completions.size(); )
+  {
+    const std::string& cmd = completions[list_index++];
+    if(cmd.compare(0, len, text) == 0)
+    {
+      return strdup(cmd.c_str());
+    }
+  }
+
+  return NULL;
+}
```

Comment:
Does this need a mutex as it modifies global (static local scope) varables ? It might be that it already can only be called from one thread, but best to ask just in case.